File: src/quickie.ml

type 'a testResult =
| Ok
| Fail of string

(* Initialize Random so it won't use the default seed *)
let _ = Random.self_init()

external toString : 'a -> string = "JSON.stringify" [@@bs.val]
external toStringProp : ('a -> bool) -> string = "String" [@@bs.new]

type rng = unit -> int

module type RNG = sig
  type state
  val make  : int -> state
  val seed : state -> int
  val int : state -> int
end

module DefaultRNG : RNG = struct
  type state = Random.State.t * int
  let make seed = (Random.State.make [| seed |], seed)
  let seed = function (_, seed) -> seed
  let int = function (r,  _) -> Random.State.int r ((1 lsl 30) - 1)
end

module Generator = struct
  type 'a t = (rng -> int -> 'a)

  let fixSize size g = fun rng _ -> g rng size

  (* sample *)
  (* recursive *)
  (* fmap *)
  (* map *)
end

module Property : sig
  type 'a t = (rng -> int -> 'a testResult)
  type 'a predicate = ('a -> bool)

  val forAll : ?message:('a -> ('a -> bool) -> string) -> 'a Generator.t -> 'a predicate -> 'a t
end = struct
  type 'a t = (rng -> int -> 'a testResult)
  type 'a predicate = ('a -> bool)

  let beta_reduce x f =
    match [%re "/function\\s*\\(([^\\)]+)\\)/"] |> Js.Re.exec f |> Js.Null.to_opt with
    | None -> f
    | Some res ->
      let arg = (res |> Js.Re.matches).(1) in
      match [%re "/function\\s+\\(.+\\)\\s*{([\\s\\S]*)}/"] |> Js.Re.exec f |> Js.Null.to_opt with
      | None -> f
      | Some res ->
        res |> Js.Re.matches
            |> (fun matches -> matches.(1))
            |> Js.String.trim
            |> Js.String.replaceByRe [%re "/^return /"] ""
            |> Js.String.replaceByRe [%re "/;$/"] ""
            |> Js.String.replaceByRe (Js.Re.fromString ("\\b" ^ arg ^ "\\b")) x

  let defaultMessage = fun sample predicate ->
    let s = toString sample in
    let p = predicate |> toStringProp |> beta_reduce s in
    s ^ " does not satisfy the property\r\n\r\n\t" ^ p

  let forAll ?(message=defaultMessage) generator predicate = fun rng size ->
    let sample = generator rng size in
    if predicate sample then Ok else Fail (message sample predicate)
end

module type Assertions = sig
  type t
  val ok : unit -> t
  val fail : string -> t
end

module Make (Assert : Assertions) : sig
  val check : 
    ?seed:int ->
    ?iterations:int ->
    'a Property.t ->
    Assert.t
end = struct
  module RNG = DefaultRNG

  let max_rng_int = (1 lsl 30) - 1

  let rec run rngState property = function
    | 0 -> Assert.ok ()
    | i -> 
      let size = max_rng_int in (* TODO *)
      let rng = fun () -> RNG.int rngState in
      match property rng size with
      | Ok ->
        run rngState property (i - 1)
      | Fail message ->
        Assert.fail ("Using seed: " ^ (string_of_int (RNG.seed rngState)) ^ "\r\n\r\n" ^ message ^ "\r\n")

  let check
    ?(seed=Random.int max_rng_int)
    ?(iterations=100)
    property
  = run (RNG.make seed) property iterations
end

include Make(struct
  type t = bool
  let ok () = true
  let fail _ = false
end)

File: lib/js/src/quickie.js
'use strict';

var Curry      = require("bs-platform/lib/js/curry.js");
var Random     = require("bs-platform/lib/js/random.js");
var Caml_array = require("bs-platform/lib/js/caml_array.js");
var Pervasives = require("bs-platform/lib/js/pervasives.js");

Random.self_init(/* () */0);

function make(seed) {
  return /* tuple */[
          Curry._1(Random.State[/* make */0], /* int array */[seed]),
          seed
        ];
}

function seed(param) {
  return param[1];
}

function $$int(param) {
  return Curry._2(Random.State[/* int */4], param[0], 1073741823);
}

var DefaultRNG = /* module */[
  /* make */make,
  /* seed */seed,
  /* int */$$int
];

function fixSize(size, g, rng, _) {
  return Curry._2(g, rng, size);
}

var Generator = /* module */[/* fixSize */fixSize];

function beta_reduce(x, f) {
  var match = (/function\s*\(([^\)]+)\)/).exec(f);
  if (match !== null) {
    var arg = Caml_array.caml_array_get(match, 1);
    var match$1 = (/function\s+\(.+\)\s*{([\s\S]*)}/).exec(f);
    if (match$1 !== null) {
      return Caml_array.caml_array_get(match$1, 1).trim().replace((/^return /), "").replace((/;$/), "").replace(new RegExp("\\b" + (arg + "\\b")), x);
    } else {
      return f;
    }
  } else {
    return f;
  }
}

function defaultMessage(sample, predicate) {
  var s = JSON.stringify(sample);
  var p = beta_reduce(s, new String(predicate));
  return s + (" does not satisfy the property\r\n\r\n\t" + p);
}

function forAll($staropt$star, generator, predicate, rng, size) {
  var message = $staropt$star ? $staropt$star[0] : defaultMessage;
  var sample = Curry._2(generator, rng, size);
  if (Curry._1(predicate, sample)) {
    return /* Ok */0;
  } else {
    return /* Fail */[Curry._2(message, sample, predicate)];
  }
}

var Property = /* module */[/* forAll */forAll];

function Make(Assert) {
  var check = function ($staropt$star, $staropt$star$1, property) {
    var seed$1 = $staropt$star ? $staropt$star[0] : Random.$$int(1073741823);
    var iterations = $staropt$star$1 ? $staropt$star$1[0] : 100;
    var rngState = make(seed$1);
    var property$1 = property;
    var iterations$1 = iterations;
    var _i = 0;
    while(true) {
      var i = _i;
      if (i >= iterations$1) {
        return Curry._1(Assert[/* ok */0], /* () */0);
      } else {
        var rng = function () {
          return $$int(rngState);
        };
        var match = Curry._2(property$1, rng, 1073741823);
        if (match) {
          return Curry._1(Assert[/* fail */1], "Iteration " + (Pervasives.string_of_int(i) + (" using seed: " + (Pervasives.string_of_int(seed(rngState)) + ("\r\n\r\n" + (match[0] + "\r\n"))))));
        } else {
          _i = i + 1 | 0;
          continue ;
          
        }
      }
    };
  };
  return /* module */[/* check */check];
}

function check($staropt$star, $staropt$star$1, property) {
  var seed$1 = $staropt$star ? $staropt$star[0] : Random.$$int(1073741823);
  var iterations = $staropt$star$1 ? $staropt$star$1[0] : 100;
  var rngState = make(seed$1);
  var property$1 = property;
  var iterations$1 = iterations;
  var _i = 0;
  while(true) {
    var i = _i;
    if (i >= iterations$1) {
      return /* true */1;
    } else {
      var rng = function () {
        return $$int(rngState);
      };
      var match = Curry._2(property$1, rng, 1073741823);
      if (match) {
        "Iteration " + (Pervasives.string_of_int(i) + (" using seed: " + (Pervasives.string_of_int(seed(rngState)) + ("\r\n\r\n" + (match[0] + "\r\n")))));
        return /* false */0;
      } else {
        _i = i + 1 | 0;
        continue ;
        
      }
    }
  };
}

exports.DefaultRNG = DefaultRNG;
exports.Generator  = Generator;
exports.Property   = Property;
exports.Make       = Make;
exports.check      = check;
/*  Not a pure module */
